File: DiningPhilosophers/Philosopher.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace DiningPhilosophers
{
    class Philosopher
    {
        public string Name { get; set; }

        public Fork LeftFork { get; set; }

        public Fork RightFork { get; set; }

        public Philosopher LeftPhilosopher { get; set; }

        public Philosopher RightPhilosopher { get; set; }

        public Waiter Waiter { get; set; }

        public int ThinkTime { get; set; }

        public int EatTime { get; set; }

        public int ThinkVariance { get; set; }

        public int EatVariance { get; set; }

        public int NextThinkTime => ThinkTime + (int)(-ThinkVariance + new Random().NextDouble() * 2 * ThinkVariance);

        public int NextEatTime => EatTime + (int)(-EatVariance + new Random().NextDouble() * 2 * EatVariance);

        public bool CanEat => !LeftFork.InUse && !RightFork.InUse;

        public void Think()
        {
            Thread.Sleep(NextThinkTime);
        }

        public async Task<bool> Eat()
        {
            //Console.WriteLine(Name + " Requested");
            await Waiter.Request(this);
            LeftFork.Take();
            //Console.WriteLine(Name + " picked up left");
            RightFork.Take();
            //EventManager.Broadcast("Eating");
            //Console.WriteLine(Name + " picked up right");
            Console.WriteLine(Name + " began eating");
            Thread.Sleep(NextEatTime);
            Console.WriteLine(Name + " finished eating");
            LeftFork.Release();
            //Console.WriteLine(Name + " put down left");
            RightFork.Release();
            //Console.WriteLine(Name + " put down right");
            EventManager.Broadcast("Finished", Name);
            return true;
        }

        public async void Run()
        {
            while (true)
            {
                await Eat();
                Think();
            }
        }
    }
}

File: go/philosopher.go
package main

import "math/rand"

import "time"

import "fmt"

import "sync"

type Philosopher struct {
	Name             string
	LeftFork         *Fork
	RightFork        *Fork
	LeftPhilosopher  *Philosopher
	RightPhilosopher *Philosopher
	Waiter           *Waiter
	ThinkTime        int
	EatTime          int
	ThinkVariance    int
	EatVariance      int
}

func getVariance(baseTime int, variance int) int {
	return baseTime + int(float32(-variance)+rand.Float32()*2.0*float32(variance))
}

func (philosopher Philosopher) NextThinkTime() int {
	return getVariance(philosopher.ThinkTime, philosopher.ThinkVariance)
}

func (philosopher Philosopher) NextEatTime() int {
	return getVariance(philosopher.EatTime, philosopher.EatVariance)
}

func (philosopher Philosopher) CanEat() bool {
	fmt.Printf("%s can eat: %t\n", philosopher.Name, !philosopher.LeftFork.InUse && !philosopher.RightFork.InUse)
	return !philosopher.LeftFork.InUse && !philosopher.RightFork.InUse
}

func (philosopher Philosopher) Think() {
	fmt.Printf("%s thinking\n", philosopher.Name)
	time.Sleep(time.Duration(philosopher.NextThinkTime()))
}

func (philosopher *Philosopher) Eat() {
	fmt.Printf("%s awaiting eat\n", philosopher.Name)
	requestChan := philosopher.Waiter.Request(philosopher)
	res := <-requestChan
	fmt.Printf("granted %t\n", res)
	philosopher.LeftFork.Take()
	philosopher.RightFork.Take()
	fmt.Printf("%s began eating\n", philosopher.Name)
	time.Sleep(time.Duration(philosopher.NextEatTime()))
	fmt.Printf("%s finished eating\n", philosopher.Name)
	philosopher.LeftFork.Release()
	philosopher.RightFork.Release()
	GetEventMangager().Broadcast("Finished", philosopher.Name)
}

func (philosopher *Philosopher) Run(wg *sync.WaitGroup) {
	defer wg.Done()
	for true {
		philosopher.Eat()
		philosopher.Think()
	}
}
